### PR TITLE
Fix for Issue #331

### LINKED ISSUE #331: Query string management is messed up
According to https://github.com/LearnBoost/socket.io-spec, multiplexed connections can have query string arguments:
```
1::/test?my=param
```

It is not the case in current implementation of the socket.io.

For example, following code:

var chat = new io.connect('http://localhost:8001/chat?a=b&c=d'),
      ping = new io.connect('http://localhost:8001/ping?d=e&f=g');

Does following:
1. Makes handshake request to the server: GET /socket.io/1/?a=b&c=d&t=1320339991081
2. Sends connection packet for chat without query string parameters: 1::/chat
3. Sends connection packet for ping without query string parameters: 1::/ping

Instead, supposed workflow should be:
1. Makes handshake request to the server: GET /socket.io/1/?t=1320339991081
2. Sends connection packet for chat with query string parameters: 1::/chat?a=b&c=d
3. Sends connection packet for ping with query string parameters: 1::/ping?d=e&f=g

Another test case can look like:
var conn = new io.connect('http://localhost:8001/?foo=bar'),
      chat = new io.connect('http://localhost:8001/chat?a=b&c=d'),
      ping = new io.connect('http://localhost:8001/ping?d=e&f=g');

and flow should be:
1. Make handshake request to the server: GET /socket.io/1/?foo=bar&t=1320339991081
2. Send connection packet for chat with query string parameters: 1::/chat?a=b&c=d
3. Send connection packet for ping with query string parameters: 1::/ping?d=e&f=g

Effectively, this means that:
1. While protocol supports query string parameters for multiplexed connections it is not being used
2. It stores parameters on socket level instead of storing parameters on connection level
3.1. As a result, first connection attempt will pass parameters
3.2. As a result, any subsequent connection attempts will lose their parameters

### PR DESCRIPTION
**Pull request by whattokingu**
Fixes #331.
Send query string over together with CONNECT packets.
Query associated with each socket is stored as a property of that socket.

### CODE REVIEW

**Comment by nkzawa** [comment]
We don't have style guide, but please follow existing styles like the following:

- `if(` => `if (`
- `for()` => `for () {}` if it's multi-line.

It would be nice to introduce `eslint`. cc @rauchg

File: test/socket.js [hunk 1]
```diff
@@ -96,4 +95,14 @@ describe('socket', function(){
       socket.compress(false).emit('hi');
     });
   });
+
+  it('should store query string as a property', function(done){
+    var socket = io('/abc', {query: {a:'b'}}); //passes in as a query obj
+    var socket2 = io('/abcd?b=c&d=e'); //passes in as a query string
+    expect(socket.query).to.be('a=b');
+    expect(socket2.query).to.be('b=c&d=e');
```
Comment (nkzawa) [inline test/socket.js]: It'd be nice if we could test if query parameter was actually sent to server?

File: lib/index.js [hunk 2]
```diff
@@ -61,8 +61,20 @@ function lookup(uri, opts) {
...
-
-  return io.socket(parsed.path);
+  if(parsed.query && !opts.query){
+    opts.query = parsed.query;
+  }else if(opts && 'object' == typeof opts.query){
+    function encodeQueryString(obj){
+      var str = [];
+      for(var p in obj)
+        if (obj.hasOwnProperty(p)) {
+          str.push(p + "=" + obj[p]);
```
Comment (nkzawa) [inline lib/index.js]: need to escape values ? (`encodeURIComponent`)

File: test/socket.js (hunk 1 above)
Comment (whattokingu) [inline test/socket.js]: The corresponding PR on [socket.io](https://github.com/socketio/socket.io/pull/2422) has the test that checks for the query parameter being sent over. Do you think that's ok?

Comment (nkzawa) [inline test/socket.js]: Then it would be ok, thanks!

File: lib/manager.js [hunk 3]
```diff
@@ -401,6 +401,7 @@ Manager.prototype.destroy = function(socket){
 Manager.prototype.packet = function(packet){
   debug('writing packet %j', packet);
   var self = this;
+  if (packet.query && packet.type == 0) packet.nsp += '?' + packet.query;
```
Comment (nkzawa) [inline lib/manager.js]: I wonder this packet can be `{ nsp: '/nsp', data:query }`. so that we don't need to construct and parse query string.

**Comment by nkzawa** [comment]
Thanks! Can you fix the conflict ?

**Comment by whattokingu** [comment]
@nkzawa Updated and fixed the conflict. Tests passed locally, including linting.

File: lib/manager.js (hunk 3 above)
Comment (nkzawa) [inline lib/manager.js]: I think `query` can be in `data` rather than appending to `nsp`. Any thoughts about that ? cc @rauchg

Comment (whattokingu) [inline lib/manager.js]: That's possible, but that would require changes to Socket.io but the corresponding [PR](https://github.com/socketio/socket.io/pull/2422) has already been merged.

**Comment by rauchg** [comment]
Thank you!

**Pull request merged**